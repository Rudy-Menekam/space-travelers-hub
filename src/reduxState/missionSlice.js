/* eslint-disable no-param-reassign */
/* eslint-disable max-len */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
// Get missions from API
export const getMission = createAsyncThunk('mission/getMission', async () => {
  const result = await axios('https://api.spacexdata.com/v3/missions');
  result.data.forEach((object) => {
    object.reserved = false;
  });
  return result.data;
});

const initialState = {
  missions: [],
  status: 'idle',
  error: null,
  reserved: [],
};

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, action) => {
      const missionIndex = state.missions.findIndex(
        (mission) => mission.mission_id === action.payload,
      );
      if (missionIndex !== -1) {
        const mission = state.missions[missionIndex];
        mission.reserved = !mission.reserved;
      }
    },
    myReservedMissions: (state) => {
      const missions = state.missions.filter(
        (mission) => mission.reserved === true,
      );
      return { ...state, reserved: missions };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMission.fulfilled, (state, action) => {
        state.missions = action.payload;
        state.status = 'succeded';
      })
      .addCase(getMission.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(getMission.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});
export const { joinMission, myReservedMissions } = missionSlice.actions;
export const allMissions = (state) => state.missions.missions;
export const missionsStatus = (state) => state.missions.status;
export const missionsError = (state) => state.missions.error;
export default missionSlice.reducer;
