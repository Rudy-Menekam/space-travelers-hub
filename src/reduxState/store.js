import { configureStore } from '@reduxjs/toolkit';
import missionReducer from './missionSlice';
import rocketReducer from './rockets/rocketSlice';

const store = configureStore({
  reducer: {
    missions: missionReducer,
    Rocket: rocketReducer,
  },
});
export default store;
