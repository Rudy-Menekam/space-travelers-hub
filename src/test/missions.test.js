import React from 'react';
import axios from 'axios';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../reduxState/store';
import { getMission, joinMission } from '../reduxState/missionSlice';
import Missions from '../components/missions';

it('Mission page renders correctly', () => {
  const tree = renderer
    .create(
      <React.StrictMode>
        <Provider store={store}>
          <BrowserRouter>
            <Missions />
          </BrowserRouter>
        </Provider>
      </React.StrictMode>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Mission redux state tests', () => {
  it('Should initially set Missions store to an empty Array', () => {
    const state = store.getState().missions;
    expect(state.missions).toEqual([]);
  });

  it('should Join Mission payload send correct', () => {
    const expectedPayload = {
      payload: '9D1B7E5',
      type: 'missions/joinMission',
    };
    const actualPayload = joinMission('9D1B7E5');
    expect(actualPayload).toEqual(expectedPayload);
  });

  it('Mission fetch data from API', async () => {
    const url = 'https://api.spacexdata.com/v3/missions';
    const axiosSpy = jest.spyOn(axios, 'get');
    jest.setTimeout(90000);
    const dispatchSpy = jest.fn();

    await getMission(url)(dispatchSpy);

    expect(axiosSpy).toHaveBeenCalledWith(url);
  });
});
