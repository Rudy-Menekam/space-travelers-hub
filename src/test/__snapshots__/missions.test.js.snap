// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission page renders correctly 1`] = `
<div
  className="missions"
>
  <div
    className="header"
  >
    <div>
      <p>
        Mission
      </p>
    </div>
    <div>
      <p>
        Description
      </p>
    </div>
    <div>
      <p>
        Status
      </p>
    </div>
    <div />
  </div>
</div>
`;
